aspnetcore/tutorials/build-your-first-blazor-app.md

```diff
@@ -1,212 +1,276 @@
+Create an *Interfaces* folder. In the folder, create an interface for the Planet Service named `IPlanetService`. An [\@using](xref:mvc/views/razor#section-1) statement for the `WebApplication.Models` namespace provides the interface with access to the `Planet` model.
 
-## Route to components
+*Interfaces/IPlanetService.cs*:
 
-The `@page` directive at the top of the *Counter.razor* file specifies that the Counter component is a routing endpoint. The Counter component handles requests sent to `/counter`. Without the `@page` directive, a component doesn't handle routed requests, but the component can still be used by other components.
+[!code-csharp[](build-your-first-blazor-app/samples_snapshot/3.x/IPlanetService.cs)]
 
-## Dependency injection
+The service requires an implementation that provides the `GetPlanets` and `AddPlanet` methods with a data source for the planets:
 
-Services registered in the app's service container are available to components via [dependency injection (DI)](xref:fundamentals/dependency-injection). Inject services into a component using the `@inject` directive.
+* For the data source, a production app would use persistent data storage, such as an [Azure Cosmos database](https://azure.microsoft.com/free/cosmos-db/) accessed by [Entity Framework (EF) Core](/ef/core/) ([in preview for Azure Cosmos](https://github.com/aspnet/EntityFrameworkCore/issues/12086)). For this tutorial experience, the app uses a `List<Planet>` stored in the app's memory.
+* In the `GetPlanets` method, the app uses [Linq](/dotnet/csharp/programming-guide/concepts/linq/) to filter planets from the planet list (`_planets`) if the `filter` has a value.
+* In the `AddPlanet` method, the app calls [ICollection\<T>.Add(T)](xref:System.Collections.Generic.ICollection`1.Add*) to add a new planet to the list.
 
-Examine the directives of the FetchData component.
+Create a *Services* folder. In the folder, create the Planet Service implementation (`PlanetService`) that implements the `IPlanetService` interface.
 
-If working with a Blazor server-side app, the `WeatherForecastService` service is registered as a [singleton](xref:fundamentals/dependency-injection#service-lifetimes), so one instance of the service is available throughout the app. The `@inject` directive is used to inject the instance of the `WeatherForecastService` service into the component.
+*Services/PlanetService.cs*:
 
-*Pages/FetchData.razor*:
+[!code-csharp[](build-your-first-blazor-app/samples_snapshot/3.x/PlanetService.cs)]
 
-[!code-cshtml[](build-your-first-blazor-app/samples_snapshot/3.x/FetchData1.razor?highlight=3)]
+Planet names are &copy;1974-1981 [BBC](https://www.bbc.co.uk/programmes/b006q2x0). Planet images are provided as a courtesy of the [National Aeronautics and Space Administration (NASA)](https://www.nasa.gov/).
```

Comment:
The snippets aren't localized. It's a move I've been making on most topics. I leave a legal notice or critical code comment in the code, too, because the sample can be separated from the text.